matching/README.md

```diff
@@ -0,0 +1,991 @@
+clause C. This ensures that a binding for the L variable in list patterns with
+variables exists by the time a rule is applied. The reason we defer this until
+the end of the decision tree is that different rows may need to construct the
+value of L differently even given the same runtime length of the list. Thus,
+we cannot simply do it as part of the specialization process, or we will
+disrupt sharing for the decision tree.
+
+```
+def foldRanges([(Occurrence, Sort, Int, Int)] R, DT dt): DT
+  case R of
+    [] -> dt
+    (l : o, s, h, t) : tail ->
+      val f = the function that removes h elements from the head of a list and t elements from its tail
+      foldRanges(tail, Function(l : o, f, [(o, s)], s, dt))
+```
+
+### Canonicalized Patterns
+
+Thus far we have not really given thought to how we define variables.
+
+In most contexts, a variable is simply a name:
+
+```
+syntax Variable ::= String
+```
+
+However, there are a few contexts where using named variables is problematic.
+In particular, in the `Constructor`, `Occurrence`, and `DT` sorts, we actually
+need to be able to reference a variable by the occurrence that bound it. Thus,
+in these contexts, we actually define a variable as follows:
+
+```
+syntax Variable ::= Occurrence
+```
+
+We thus have two notions of patterns: patterns whose variables are names, and
+patterns whose variables are occurrences. As a result, we need some way of
+converting between these representations.
+
+As it happens, the clause list of a matrix provides this information in the
+form of the substitution S. We thus say that to *canonicalize* a pattern p
+with respect to row i of a Matrix m is to substitute all named variables in p
+with their bindings in S_i. We say that a pattern is *bound* if all variables
+in p have a binding in S_i.
+
+### Map Patterns
+
+K does not implement full AC matching on maps or sets. Instead, it attempts to
+order the matching process in such a manner that wherever possible, the
+substitution for all variables in the keys of map patterns are bound before
+matching on the map itself. This allows us to, in the vast majority of cases,
+compile matching on maps to map membership and lookup checks, which are vastly
+more efficient than AC matching.
+
+As a result, K has two types of matching on map patterns. The first, when the
+key of the map element being matched on is bound, is called a *lookup*. In the
+event that the pattern is not bound when matching occurs, we call this a
+*map choice*. Essentially, the intuition behind a map choice is that we try
+each element of the map in turn, proceeding as if it is the correct one, until
+matching fails, at which point we backtrack to the next map element. In the
+worst case, if matching a rule requires n map choices on a map of size m, then
+the complexity of this algorithm is `O(m^n)`. This is, of course, worse than
+the complexity of linear AC matching, which is `O(s*t^3)` where s is the size
```

Comment:
Do we have nice citations for AC matching? If we're claiming this is "of course" we should cite the AC complexity